### DISCUSSION TITLE
Additional invoice template variables needed

### DISCUSSION THREAD

**Discussion by vic-t** [kimai/kimai#2351] [General]
I love the fact that Kimai allows the creation of Invoices in Word format which allows for simple last minute changes of layout and content. 

However, I feel a bit restricted by the layout possibilities Kimai offers built-in. I understand that I could probably just switch to Twig templates and have a lot more possibilities but I so much prefer the simplicity of Word templates. 

For example, I would like to work with an invoice that, in its first section, shows all times and efforts in a summarized way, for example like this: 

Timesheet total: 10 hours, EUR 500 (maybe even broken down by user or activity)
Fixed rate total: EUR 400 (maybe broken down by activity)
Expenses: EUR 150
Subtotal
VAT
Total

And then, in the next section, I'd create a list of all detailed entries which is possible already today. 

I think in the end what I'm looking for at this stage are additional variables that represent the information mentioned above. If I were to give them names, I'd probably call them: 

```
${invoice.timesheet_total}
${invoice.timesheet_user1_total}  
${invoice.timesheet_user2_total}
${invoice.fixedrates_total}
${invoice.activity1_total}
${invoice.activity2_total}
${invoice.expenses_type1_total} # Types here would stand for the different types of expenses possible
${invoice.expenses_type2_total} # Types here would stand for the different types of expenses possible
```
I understand this is not as easily done as it looks, particularly since both fixed rates as well as expenses are basically also just timesheet entries. But I believe some sort of categorization is present and therefore it should be possible to summarize these entries. 

I intentionally didn't create a feature request for now because there may be other, better ways of achieving the same goal so I'm just putting this up for discussion.

**Comment by kevinpapst** [comment]
You can write new invoice hydrators, see existing ones: https://github.com/kevinpapst/kimai2/tree/master/src/Invoice/Hydrator

These hydrator objects extract information pieces from the InvoiceModel and registers them as new variables, which could then be used in your invoice template.
